scapy/contrib/automotive/obd/scanner.py

```diff
@@ -0,0 +1,144 @@
+    supported_ids_req = OBD()/service_class(b'\x00')
+
+    while supported_ids_req is not None:
+        resp = socket.sr1(supported_ids_req, timeout=timeout, verbose=verbose)
+
+        # If None, the device did not respond.
+        # Usually only occurs, if device is off.
+        if resp is None:
+            break
+
+        supported_ids_req = None
+
+        all_supported_in_range = getattr(resp.data_records[0], supported_prop)
+
+        for supported in all_supported_in_range:
+            id_number = int(supported[-2:], 16)
+            supported_id_numbers.add(id_number)
+
+            # check whether supported id is for next id range
+            if id_number % 20 == 0:
+                supported_ids_req = OBD()/service_class(chb(id_number))
+
+    return supported_id_numbers
+
+
+def _scan_id_service(socket, timeout, service_class, id_numbers, verbose):
+    data = dict()
+
+    for id_number in id_numbers:
+        id_byte = chb(id_number)
+        # assemble request packet
+        pkt = OBD()/service_class(id_byte)
+        resp = socket.sr1(pkt, timeout=timeout, verbose=verbose)
+
+        if resp is not None:
+            data[id_number] = bytes(resp)
+    return data
+
+
+def _scan_dtc_service(socket, timeout, service_class, verbose):
+    req = OBD()/service_class()
+    resp = socket.sr1(req, timeout=timeout, verbose=verbose)
+    if resp is not None:
+        return bytes(resp)
+
+
+def obd_scan(socket, timeout=0.1, supported_ids=False, unsupported_ids=False, verbose=False):
+    dtc = dict()
+    supported = dict()
+    unsupported = dict()
+
+    print("\nStarting OBD-Scan...")
+
+    print("\nScanning Diagnostic Trouble Codes:")
+    # Emission-related DTCs
+    dtc[3] = _scan_dtc_service(socket, timeout, OBD_S03, verbose)
+    # Emission-related DTCs detected during current or last completed driving cycle
+    dtc[7] = _scan_dtc_service(socket, timeout, OBD_S07, verbose)
+    # Permanent DTCs
+    dtc[10] = _scan_dtc_service(socket, timeout, OBD_S0A, verbose)
+    print("Service 3:")
+    print(dtc[3])
+    print("Service 7:")
+    print(dtc[7])
+    print("Service 10:")
+    print(dtc[10])
+
+    if not supported_ids and not unsupported_ids:
+        return dtc
+
+    # Powertrain
+    supported_ids_s01 = _supported_id_numbers(socket, timeout, OBD_S01, 'pid', verbose)
+    # On-board monitoring testresults for non-continuously monitored systems
+    supported_ids_s06 = _supported_id_numbers(socket, timeout, OBD_S06, 'mid', verbose)
+    # Control of on-board system, test or component
+    supported_ids_s08 = _supported_id_numbers(socket, timeout, OBD_S08, 'tid', verbose)
+    # On-board monitoring test results for non-continuously monitored systems
+    supported_ids_s09 = _supported_id_numbers(socket, timeout, OBD_S09, 'iid', verbose)
+
+    if supported_ids:
+        print("\nScanning supported Parameter IDs")
+        supported[1] = _scan_id_service(socket, timeout, OBD_S01, supported_ids_s01, verbose)
```

Comment:
What about `OBD_S02`?

Reply:
On consultation predecessor @akorb stated, that this service, which is being used to monitor freeze frames, is hardly being used, which is why, for us it didn't seem to be worth the effort.

What is your opinion about that? 
Should it be implemented as well?

Reply:
no, it's fine